File: src/main/java/com/qtech/bubbles/init/Entities.java
package com.qtech.bubbles.init;

import com.qtech.bubbles.QInternalAddon;
import com.qtech.bubbles.common.gametype.AbstractGameType;
import com.qtech.bubbles.common.init.ObjectInit;
import com.qtech.bubbles.common.screen.Screen;
import com.qtech.bubbles.entity.AmmoEntity;
import com.qtech.bubbles.entity.BubbleEntity;
import com.qtech.bubbles.entity.GiantBubbleEntity;
import com.qtech.bubbles.entity.player.PlayerEntity;
import com.qtech.bubbles.entity.types.EntityType;
import com.qtech.bubbles.registry.DeferredRegister;
import com.qtech.bubbles.registry.Registers;
import com.qtech.bubbles.registry.object.RegistryObject;

import java.util.function.Supplier;

@SuppressWarnings("ALL")
//@ObjectHolder(addonId = "qbubbles")
public class Entities implements ObjectInit<EntityType> {
    public static final DeferredRegister<EntityType> ENTITIES = DeferredRegister.create(QInternalAddon.ADDON_ID, Registers.ENTITIES);

    public static final RegistryObject<EntityType<AmmoEntity>> AMMO = register("bubble", () -> new EntityType<>(AmmoEntity::new));
    public static final RegistryObject<EntityType<BubbleEntity>> BUBBLE = register("bubble", () -> new EntityType<>(BubbleEntity::new));
    public static final RegistryObject<EntityType<GiantBubbleEntity>> GIANT_BUBBLE = register("giant_bubble", () -> new EntityType<>(GiantBubbleEntity::new));
    public static final RegistryObject<EntityType<PlayerEntity>> PLAYER = register("entity", () -> new EntityType<>(PlayerEntity::new));

    @SuppressWarnings("SameParameterValue")
    private static <T extends EntityType<?>> RegistryObject<T> register(String name, Supplier<T> supplier) {
        return ENTITIES.register(name, supplier);
    }
}

File: src/main/kotlin/com/qtech/bubbles/init/Entities.kt
package com.qtech.bubbles.init

import com.qtech.bubbles.BBInternalAddon
import com.qtech.bubbles.common.gametype.AbstractGameMode
import com.qtech.bubbles.common.init.ObjectInit
import com.qtech.bubbles.entity.AmmoEntity
import com.qtech.bubbles.entity.BubbleEntity
import com.qtech.bubbles.entity.GiantBubbleEntity
import com.qtech.bubbles.entity.player.PlayerEntity
import com.qtech.bubbles.entity.types.EntityType
import com.qtech.bubbles.registry.DeferredRegister
import com.qtech.bubbles.registry.Registers
import com.qtech.bubbles.registry.`object`.RegistryObject
import java.util.function.Supplier

//@ObjectHolder(addonId = "qbubbles")
object Entities : ObjectInit<EntityType<*>?> {
    val ENTITIES: DeferredRegister<EntityType<*>> = DeferredRegister.create(BBInternalAddon.ADDON_ID, Registers.ENTITIES)
    val AMMO: RegistryObject<EntityType<AmmoEntity>> = register("bubble") { EntityType { gameMode: AbstractGameMode -> AmmoEntity(gameMode) } }
    val BUBBLE: RegistryObject<EntityType<BubbleEntity>> = register("bubble") { EntityType { gameMode: AbstractGameMode -> BubbleEntity(gameMode) } }
    val GIANT_BUBBLE: RegistryObject<EntityType<GiantBubbleEntity>> = register("giant_bubble") { EntityType { gameMode: AbstractGameMode -> GiantBubbleEntity(gameMode) } }
    val PLAYER: RegistryObject<EntityType<PlayerEntity>> = register("entity") { EntityType { gameType -> PlayerEntity(gameType) } }
    private fun <T : EntityType<*>> register(name: String, supplier: Supplier<T>): RegistryObject<T> {
        return ENTITIES.register(name, supplier)
    }
}